### PR TITLE
fix(admin): add confirmation modal for user role change and user (un)deletion

### LINKED ISSUE #2481: 🐛 [Bug] [Admin] No confirmation when changing user role
## 🐛 Bugreport

The user is not asked if he wants to change a users role. We should ask the user e.g. via alert window if he really wants to change a users role in the admin interface.

### PR DESCRIPTION
**Pull request by mahula**
## 🍰 Pullrequest
In the admin interface these actions require proper confirmation modals:
- changing a user's role
- (un)deleting a user

### Issues
- fixes #2481 

### Todo
- [X] add confirmation flow to user (un)deletion actions
  - [x] remove toggle and persent the `delete` / `undelete` Button
         delete_user-start
         undelete_user-start

  - [x] add confirmation modal for user (un)deletion actions
         delete_user-overlay
        undelete_user-overlay

- [x] add confirmation flow to user role change
  - [x] disabled action button, if role is not changed
         change_user_role-start_without_change

  - [x] enabeld action button, if role is changed
        change_user_role-start_with_role_changed
  - [x] add confirmation modal for user role change actions
        user_role_confirm_modal

### CODE REVIEW

File: admin/src/components/DeletedUserFormular.vue [hunk 1]
```diff
@@ -4,19 +4,36 @@
...
-        <div>{{ item.deletedAt ? $t('undelete_user') : $t('delete_user') }}</div>
-      </b-form-checkbox>
-
       <div class="mt-3 mb-5">
-        <b-button v-if="checked && item.deletedAt === null" variant="danger" @click="deleteUser">
+        <b-button
+          v-if="item.deletedAt === null"
+          variant="danger"
+          v-b-modal.delete-user-modal
+          @click="showModal('deleteUser')"
```
Comment (ulfgebhardt) [inline admin/src/components/DeletedUserFormular.vue:12]: What you use `v-b-modal` for when you reference the modal by a parameter in `showModal('deleteUser')` ?
Am I missing something?

Reference: https://stackoverflow.com/questions/49123108/vuejs-how-to-use-v-b-modal-directive-in-bootstrapvue

```
 methods: {
    showModal() {
      this.$refs.myModalRef.show();
    },
  }
  ```

File: admin/src/components/DeletedUserFormular.vue [hunk 2]
```diff
@@ -4,19 +4,36 @@
...
+        >
           {{ $t('delete_user') }}
         </b-button>
-        <b-button v-if="checked && item.deletedAt !== null" variant="success" @click="unDeleteUser">
+        <b-button
+          v-if="item.deletedAt !== null"
+          variant="success"
+          v-b-modal.delete-user-modal
+          @click="showModal('undeleteUser')"
```
Comment (ulfgebhardt) [inline admin/src/components/DeletedUserFormular.vue:20]: same

File: admin/src/components/DeletedUserFormular.vue [hunk 3]
```diff
@@ -34,9 +51,34 @@ export default {
...
+      }
+
+      if (type === 'undeleteUser') {
+        this.variant = 'success'
+        this.modalTitle = this.$t('overlay.undeleteUser.title')
+        this.modalQuestion = this.$t('overlay.undeleteUser.question', { username: this.username })
+        this.modalOkTitle = this.$t('overlay.undeleteUser.yes')
+        this.modalEvent = this.unDeleteUser
+      }
+    },
```
Comment (ulfgebhardt) [inline admin/src/components/DeletedUserFormular.vue:81]: I have no further insight into this - but for me this looks like you just want to define two modals, each with the correct values submitted and open them accordingly.

The whole mechanic behind `v-b-modal` seems to support my statement here I think - it seems like the whole mechanic is made to allow to define multiple modals which then open when you call `showModal` with the specific reference defined.
Reference: https://bootstrap-vue.org/docs/components/modal#using-v-b-modal-directive

And here is a multi-modal example in the docu: https://bootstrap-vue.org/docs/components/modal#multiple-modal-support

File: admin/src/locales/de.json [hunk 4]
```diff
@@ -126,11 +127,26 @@
       "title": "Gemeinwohl-Beitrag löschen!",
       "yes": "Ja, Beitrag löschen!"
     },
+    "deleteUser": {
+      "question": "Willst du {username} wirklich löschen?",
+      "title": "Nutzer löschen",
+      "yes": "Ja, Nutzer löschen"
```
Comment (ulfgebhardt) [inline admin/src/locales/de.json:133]: why not use `ok` as it is defined in the modal?

File: admin/src/locales/de.json [hunk 5]
```diff
@@ -126,11 +127,26 @@
...
     "deny": {
       "question": "Willst du diesen Gemeinwohl-Beitrag  wirklich ablehnen?",
       "text": "Nach dem Speichern ist der Datensatz nicht mehr änderbar und kann auch nicht mehr gelöscht werden. Bitte überprüfe genau, dass alles stimmt.",
       "title": "Gemeinwohl-Beitrag ablehnen!",
       "yes": "Ja, Beitrag ablehnen und speichern!"
+    },
+    "undeleteUser": {
+      "question": "Willst du wirklich {username} wiederherstellen?",
+      "title": "Nutzer wiederherstellen",
+      "yes": "Ja, Nutzer wiederherstellen"
```
Comment (ulfgebhardt) [inline admin/src/locales/de.json:144]: same

File: admin/src/locales/de.json [hunk 6]
```diff
@@ -126,11 +127,26 @@
...
+    },
+    "changeUserRole": {
+      "question": "Willst du die Rolle von {username} wirklich zu {newRole} ändern?",
+      "title": "Nutzerrolle ändern",
+      "yes": "Ja, Nutzerrolle ändern"
```
Comment (ulfgebhardt) [inline admin/src/locales/de.json:149]: same

File: admin/src/locales/en.json [hunk 7]
```diff
@@ -126,11 +127,26 @@
       "title": "Delete creation!",
       "yes": "Yes, delete and save creation!"
     },
+    "deleteUser": {
+      "question": "Do you really want to delete {username}?",
+      "title": "Delete user",
+      "yes": "Yes, delete user"
```
Comment (ulfgebhardt) [inline admin/src/locales/en.json:133]: same

File: admin/src/locales/de.json [hunk 8]
```diff
@@ -126,11 +127,26 @@
       "title": "Gemeinwohl-Beitrag löschen!",
       "yes": "Ja, Beitrag löschen!"
     },
+    "deleteUser": {
```
Comment (ulfgebhardt) [inline admin/src/locales/de.json:130]: We should think about grouping this into
```
modals: {
  modalA:{
    question: ...
  },
  modalB: {..}
}
```
no?

**Comment by ulfgebhardt** [comment]
image
image

Furthermore - and this is clearly not part of this PR - I think it could be wise to group this into one tab `Nutzer-Verwaltung`

File: admin/src/components/ChangeUserRoleFormular.vue [hunk 9]
```diff
@@ -35,21 +51,28 @@ export default {
...
-      }
-    },
-  },
   methods: {
+    showModal() {
+      this.username = `${this.item.firstName} ${this.item.lastName}`
+      this.modalTitle = this.$t('overlay.changeUserRole.title')
+      this.modalQuestion = this.$t('overlay.changeUserRole.question', {
+        username: this.username,
+        newRole: this.roleSelected, // TODO get the text:Administrator instead of admin
```
[the remark this answers is not in the document]
Comment (mahula) [reply admin/src/components/ChangeUserRoleFormular.vue:72]: This is WIP, which will be handled in this PR, and that small TODO does not need a separate issue.

File: admin/src/locales/de.json (hunk 6 above)
Comment (mahula) [reply admin/src/locales/de.json:149]: For the open contribution table's overlay these more detailed texts for the ok button where introduced to clearly show the admin, what action is behind the buton.

The modals in this PR are meant to adapt to this informative button text approach - for coherence in the admin interface.

File: admin/src/locales/en.json (hunk 7 above)
Comment (mahula) [reply admin/src/locales/en.json:133]: For the open contribution table's overlay these more detailed texts for the ok button where introduced to clearly show the admin, what action is behind the buton.

The modals in this PR are meant to adapt to this informative button text approach - for coherence in the admin interface.

File: admin/src/components/ChangeUserRoleFormular.vue [hunk 10]
```diff
@@ -35,21 +51,28 @@ export default {
...
+      currentRole: this.item.isAdmin ? rolesValues.admin : rolesValues.user,
       roleSelected: this.item.isAdmin ? rolesValues.admin : rolesValues.user,
       roles: [
         { value: rolesValues.user, text: this.$t('userRole.selectRoles.user') },
         { value: rolesValues.admin, text: this.$t('userRole.selectRoles.admin') },
       ],
+      modalTitle: '',
+      modalQuestion: '',
+      modalOkTitle: '',
+      username: '',
```
[the remark this answers is not in the document]
Comment (mahula) [reply admin/src/components/ChangeUserRoleFormular.vue]: Resolved in https://github.com/gradido/gradido/pull/2880/commits/17f7e8adf0ce2928c5b4f3bb9a1c3c13c69066a7

File: admin/src/components/DeletedUserFormular.vue (hunk 1 above)
Comment (mahula) [reply admin/src/components/DeletedUserFormular.vue]: Resolved in https://github.com/gradido/gradido/pull/2880/commits/17f7e8adf0ce2928c5b4f3bb9a1c3c13c69066a7

File: admin/src/locales/de.json (hunk 8 above)
Comment (mahula) [reply admin/src/locales/de.json:130]: It is grouped under `overlay` in the locales.

File: admin/src/components/DeletedUserFormular.vue (hunk 3 above)
Comment (mahula) [reply admin/src/components/DeletedUserFormular.vue]: This is resolved by utilizing the `bvModal.msgBoxConfirm()` method.

File: admin/src/components/DeletedUserFormular.vue (hunk 2 above)
Comment (mahula) [reply admin/src/components/DeletedUserFormular.vue]: Resolved

File: admin/src/components/ChangeUserRoleFormular.spec.js [hunk 11]
```diff
@@ -70,12 +71,16 @@ describe('ChangeUserRoleFormular', () => {
...
+        expect(wrapper.find('select.role-select').exists()).toBe(false)
+      })
+
+      it('has no button', () => {
+        expect(wrapper.find('button.btn.btn-dange').exists()).toBe(false)
       })
     })
 
-    describe('change others role', () => {
+    describe("change other user's role", () => {
```
Comment (Elweyn) [inline admin/src/components/ChangeUserRoleFormular.spec.js:83]: Prefer '' to ""
```suggestion
    describe('change other user\'s role', () => {
```

File: admin/src/components/DeletedUserFormular.vue [hunk 12]
```diff
@@ -4,15 +4,21 @@
...
     </div>
     <div v-else class="mt-5">
-      <b-form-checkbox switch size="lg" v-model="checked">
-        <div>{{ item.deletedAt ? $t('undelete_user') : $t('delete_user') }}</div>
-      </b-form-checkbox>
-
       <div class="mt-3 mb-5">
-        <b-button v-if="checked && item.deletedAt === null" variant="danger" @click="deleteUser">
+        <b-button
+          v-if="item.deletedAt === null"
```
Comment (Elweyn) [inline admin/src/components/DeletedUserFormular.vue:9]: Not certain if necessary but what happen if deletedAt is undefined? Following suggestion catches deletedAt = null, deletedAt = undefined and also deletedAt = 0.
```suggestion
          v-if="!item.deletedAt"
```

File: admin/src/components/DeletedUserFormular.vue [hunk 13]
```diff
@@ -4,15 +4,21 @@
...
+          variant="danger"
+          v-b-modal.delete-user-modal
+          @click="showDeleteModal()"
+        >
           {{ $t('delete_user') }}
         </b-button>
-        <b-button v-if="checked && item.deletedAt !== null" variant="success" @click="unDeleteUser">
+        <b-button
+          v-if="item.deletedAt !== null"
```
Comment (Elweyn) [inline admin/src/components/DeletedUserFormular.vue:17]: You check the opposite in the previous block
```suggestion
          v-else
```

**Review by Elweyn — CHANGES_REQUESTED** [review]
Nice PR, I have some small changes requested. Look at them and judge if you want to add them or discard them.

**Comment by Elweyn** [comment]
Sure we can take the suggestions that @ulfgebhardt gave in it too!

**Review by Elweyn — APPROVED** [review]

**Pull request merged**